File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.js
/**
 * // Definition for a _Node.
 * function _Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {_Node} root
 * @return {_Node}
 */
var connect = function(root) {
    if (!root) return null

    const queue = [root]

    while (queue.length > 0) {
        const levelSize = queue.length

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()

            if (i < levelSize - 1) {
                node.next = queue[0]
            }

            if (node.left) queue.push(node.left)
            if (node.right) queue.push(node.right)
        }
    }

    return root
};

File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None

        leftmost = root

        while leftmost.left:
            current = leftmost
            
            while current:

                if current.right:
                    current.left.next = current.right

                if current.next:
                    current.right.next = current.next.left
                else:
                    current.right.next = None

                current = current.next

            leftmost = leftmost.left
        
        return root